File: stock_quotation.py
from flask import Flask, request, jsonify
import requests
import time
import base64
from functools import wraps

# 请求参数 ?code=xxxx&name=xxxx 或者 ?index=xxx
app = Flask(__name__)

stock_codes = [
    {"code": "603456", "name": "九洲药业"},
    {"code": "002557", "name": "洽洽食品"},
    {"code": "002415", "name": "海康威视"},
]

auth_username = "username"
auth_password = "password"

def fetch_data(code, name=""):
    url = "https://finance.pae.baidu.com/selfselect/getstockquotation"
    params = {
        "all": "1",
        "code": code,
        "isStock": "true",
        "newFormat": "1",
        "group": "quotation_minute_ab"
    }
    
    cookieString = "your cookie string";
    
    headers = {
        'accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7',
        'accept-language': 'zh-CN,zh;q=0.9',
        'cache-control': 'max-age=0',
        'priority': 'u=0, i',
        'upgrade-insecure-requests': '1',
        'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/137.0.0.0 Safari/537.36',
        'cookie': cookieString
    }
    speed_time = time.time()

    resp = requests.get(url, headers=headers, params=params)
    resp.raise_for_status()

    data = resp.json()

    chart_point_length = 0
    # 提取 price 数组
    priceinfo = data["Result"]["priceinfo"]
    #print(len(priceinfo))
    if not priceinfo or len(priceinfo) == 0:
        minbar = []
        chart_point_length = 0
    elif len(priceinfo) <= 121: #上午
        minbar = [item["price"] for item in priceinfo[::2]]
        chart_point_length = 60
    else: #下午
        minbar = [item["price"] for item in priceinfo[::4]]
        chart_point_length = 60

    # 提取 pankouinfos.origin_pankou
    pankou = data["Result"].get("pankouinfos", {}).get("origin_pankou", {})

    price_datetime = data["Result"]["priceinfo"][-1]["datetime"]
    stock_market_code = data["Result"]["basicinfos"]["stock_market_code"]
    pankou["updateDate"] = price_datetime.split(" ")[0]
    pankou["updateTime"] = price_datetime.split(" ")[1]
    pankou["stockCode"] = stock_market_code if not name else (stock_market_code[0:6] + " " + name)

    preClose = float(pankou["preClose"])
    currentPrice = float(pankou["currentPrice"])
    high = float(pankou["high"])
    low = float(pankou["low"])

    if not high or high == 0:
        high = preClose * 1.1
        chart_max_per = 9.99
    else:
        chart_max_per = (high - preClose) * 100 / preClose

    if not low or low == 0:
        low = preClose * 0.9
        chart_min_per = -9.99
    else:
        chart_min_per = (low - preClose) * 100 / preClose

    top_icon_2 = ""
    if currentPrice:
        if currentPrice < preClose: # 绿色
            top_icon_2 = 1
        elif currentPrice > preClose: # 红色
            top_icon_2 = 2

    speed_time = time.time() - speed_time
    return {
        "chart_info": {
            "top_label_1": f"{currentPrice:.2f}",       #左上
            "top_icon_2": top_icon_2,                   #中上图标
            "top_label_2": pankou["priceLimit"] + " %", #右上
            "chart_max_val": f"{high:.2f}",             #表格左上
            "chart_min_val": f"{low:.2f}",              #表格左下
            "chart_max_per": f"{chart_max_per:.2f}" + " %", #表格右上
            "chart_min_per": f"{chart_min_per:.2f}" + " %", #表格右下
            "chart_abscissa": f"{preClose:.2f}",        #表格横轴
            "bottom_label_1": pankou["stockCode"],      #左下
            "bottom_label_2": pankou["updateDate"],     #中下
            "bottom_label_3": pankou["updateTime"],     #右下
            "chart_point_length": max(chart_point_length, len(minbar)),   #表格横轴点数
            "chart_array_length": len(minbar),          #数组长度（用不到）
        },
        "chart_array": minbar,
        "speed_time": f"{speed_time:.2f}"
    }
    
    
# HTTP 认证装饰器
def requires_auth(f):
    @wraps(f)
    def decorated(*args, **kwargs):
        auth = request.headers.get("Authorization")
        if not auth or not auth.startswith("Basic "):
            return (
                jsonify({"message": "Authentication required"}),
                401,
                {"WWW-Authenticate": "Basic realm='Login Required'"},
            )

        try:
            encoded_credentials = auth.split(" ")[1]
            decoded_credentials = base64.b64decode(encoded_credentials).decode("utf-8")
            username, password = decoded_credentials.split(":", 1)
        except Exception as e:
            print(f"Error decoding credentials: {e}")
            return (
                jsonify({"message": "Invalid authentication credentials"}),
                401,
                {"WWW-Authenticate": "Basic realm='Login Required'"},
            )

        # 验证用户名和密码
        if username == auth_username and password == auth_password:
            print("Authentication successful!")
            return f(*args, **kwargs)
        else:
            print(f"Authentication failed for user: {username}")
            return (
                jsonify({"message": "Invalid credentials"}),
                401,
                {"WWW-Authenticate": "Basic realm='Login Required'"},
            )

    return decorated


@app.route("/favicon.ico")
def favicon():
    return f"Request Path: {request.path}"


index = 0
@app.route("/getChartInfo", methods=["GET"])
# @requires_auth  # 应用 HTTP 认证装饰器
def get_minbar():
    code = request.args.get("code")
    name = request.args.get("name")
    indexStr = request.args.get("index")
    if not code:
        global index
        if indexStr and indexStr.isdigit():
            index = int(indexStr) % len(stock_codes)
        else:
            index += 1
            if index >= len(stock_codes):
                index = 0
        code = stock_codes[index]["code"]
        name = stock_codes[index]["name"]
    try:
        result = fetch_data(code, name)
        return jsonify(result)
    except Exception as e:
        return jsonify({"error": str(e)}), 500

if __name__ == "__main__":
    app.run(host="0.0.0.0", port=5000, debug=False)

File: stock_quotation.js

// cloudflare pages 部署
export default {
  async fetch(request, env, ctx) {
    const url = new URL(request.url);
    const params = url.searchParams;

    // 获取请求参数 ?code=xxxx&name=xxxx 或者 ?index=xxx
    const indexStr = params.get("index");
    let code = params.get("code");
    let name = params.get("name") || "";

    if (!code && !indexStr) {
      return new Response(JSON.stringify({ error: "Missing 'code' or 'index' parameter" }), {
        status: 400,
        headers: { "content-type": "application/json" },
      });
    }

    const stock_codes = [
        {"code": "603456", "name": "九洲药业"},
        {"code": "002557", "name": "洽洽食品"},
        {"code": "002415", "name": "海康威视"},
    ]

    if (!code) {
      if (Number.isNaN(indexStr)) {
        return new Response(JSON.stringify({ error: "'index' not a number" }), {
          status: 400,
          headers: { "content-type": "application/json" },
        });
      }
      const index = parseInt(indexStr) % stock_codes.length;
      const stock = stock_codes[index];
      code = stock.code;
      name = stock.name;
    }

    // 记录开始时间
    const startTime = Date.now();

    // 构造百度 API URL
    const targetUrl = new URL("https://finance.pae.baidu.com/selfselect/getstockquotation");
    targetUrl.searchParams.append("all", "1");
    targetUrl.searchParams.append("code", code);
    targetUrl.searchParams.append("isStock", "true");
    targetUrl.searchParams.append("newFormat", "1");
    targetUrl.searchParams.append("group", "quotation_minute_ab");

    const cookieString = "your cookie string";

    const headers = {
      'accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7',
      'accept-language': 'zh-CN,zh;q=0.9',
      'cache-control': 'max-age=0',
      'priority': 'u=0, i',
      'upgrade-insecure-requests': '1',
      'origin': 'https://gushitong.baidu.com',
      'referer': 'https://gushitong.baidu.com',
      'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/137.0.0.0 Safari/537.36',
      'cookie': cookieString
    };

    try {
      const resp = await fetch(targetUrl.toString(), {
        method: "GET",
        headers: headers
      });

      if (!resp.ok) {
        throw new Error(`Upstream API error: ${resp.status}`);
      }

      const data = await resp.json();

      // === 数据处理逻辑 (对应 Python) ===
      
      let chart_point_length = 0;
      let minbar = [];
      
      const priceinfo = data.Result.priceinfo || [];
      
      if (priceinfo.length === 0) {
        minbar = [];
        chart_point_length = 0;
      } else if (priceinfo.length <= 121) {
        // 上午: Python [::2] 等同于 index % 2 === 0
        minbar = priceinfo.filter((_, i) => i % 2 === 0).map(item => item.price);
        chart_point_length = 60;
      } else {
        // 下午: Python [::4] 等同于 index % 4 === 0
        minbar = priceinfo.filter((_, i) => i % 4 === 0).map(item => item.price);
        chart_point_length = 60;
      }

      const pankou = data.Result.pankouinfos?.origin_pankou || {};
      const price_datetime = priceinfo[priceinfo.length - 1]?.datetime || "";
      const stock_market_code = data.Result.basicinfos?.stock_market_code || "";

      // 日期格式处理
      const [datePart, timePart] = price_datetime.split(" ");
      pankou.updateDate = datePart?.replace(/-/g, ".") || "";
      pankou.updateTime = timePart?.replace(/:/g, ".") || "";
      pankou.stockCode = name ? (stock_market_code.substring(0, 7) + name) : stock_market_code;

      const preClose = parseFloat(pankou.preClose);
      const currentPrice = parseFloat(pankou.currentPrice);
      let high = parseFloat(pankou.high);
      let low = parseFloat(pankou.low);

      let chart_max_per = 0.00;
      let chart_min_per = 0.00;

      if (!high || high === 0) {
        high = preClose * 1.1;
        chart_max_per = 9.99;
      } else {
        chart_max_per = (high - preClose) * 100 / preClose;
      }

      if (!low || low === 0) {
        low = preClose * 0.9;
        chart_min_per = -9.99;
      } else {
        chart_min_per = (low - preClose) * 100 / preClose;
      }

      let top_icon_2 = "";
      if (currentPrice) {
        if (currentPrice < preClose) top_icon_2 = "1"; // 绿色
        else if (currentPrice > preClose) top_icon_2 = "2"; // 红色
      }

      const speedTimeVal = (Date.now() - startTime) / 1000;

      const result = {
        "chart_info": {
          "top_label_1": currentPrice.toFixed(2),
          "top_icon_2": top_icon_2,
          "top_label_2": pankou.priceLimit + " %",
          "chart_max_val": high.toFixed(2),
          "chart_min_val": low.toFixed(2),
          "chart_max_per": chart_max_per.toFixed(2) + " %",
          "chart_min_per": chart_min_per.toFixed(2) + " %",
          "chart_abscissa": preClose.toFixed(2),
          "bottom_label_1": pankou.stockCode,
          "bottom_label_2": pankou.updateDate,
          "bottom_label_3": pankou.updateTime,
          "chart_point_length": Math.max(chart_point_length, minbar.length),
          "chart_array_length": minbar.length,
        },
        "chart_array": minbar,
        "speed_time": speedTimeVal.toFixed(2)
      };

      return new Response(JSON.stringify(result), {
        headers: {
          "content-type": "application/json;charset=UTF-8",
          "Access-Control-Allow-Origin": "*" // 允许跨域调用
        }
      });

    } catch (err) {
      return new Response(JSON.stringify({ error: err.message }), {
        status: 500,
        headers: { "content-type": "application/json" }
      });
    }
  }
};